miri-script/src/commands.rs

```diff
@@ -0,0 +1,652 @@
+            cmd!(sh, "cargo +{toolchain} install {extra_flags...} --path {path} --force --root {sysroot} {args...}").run()?;
+            Ok(())
+        }
+        let sh = self.shell_with_miri_env()?;
+        install_to_dir(
+            &sh,
+            self.miri_dir.as_ref(),
+            &self.active_toolchain.as_ref(),
+            self.cargo_extra_flags.as_ref(),
+            &self.extra_vars.sysroot.as_ref(),
+            flags,
+        )?;
+        let cargo_miri_dir = path!(self.miri_dir / "cargo-miri");
+        install_to_dir(
+            &sh,
+            cargo_miri_dir.as_ref(),
+            &self.active_toolchain.as_ref(),
+            self.cargo_extra_flags.as_ref(),
+            self.extra_vars.sysroot.as_ref(),
+            flags,
+        )?;
+        Ok(())
+    }
+
+    fn build(&self, flags: &[OsString]) -> Result<()> {
+        // Build, and let caller control flags.
+        let toolchain = self.get_toolchain_override();
+        let miri_manifest = path!(self.miri_dir / "Cargo.toml");
+        let cargo_miri_manifest = path!(self.miri_dir / "cargo-miri" / "Cargo.toml");
+        Self::build_package(miri_manifest.as_ref(), &toolchain, &self.cargo_extra_flags, flags)?;
+        Self::build_package(
+            cargo_miri_manifest.as_ref(),
+            &toolchain,
+            &self.cargo_extra_flags,
+            flags,
+        )?;
+        Ok(())
+    }
+
+    fn check(&self, flags: &[OsString]) -> Result<()> {
+        fn check_package(
+            // Path to Cargo.toml file of a package to check.
+            path: &OsStr,
+            toolchain: &OsStr,
+            extra_flags: &[String],
+            all_targets: bool,
+            args: impl IntoIterator<Item = impl AsRef<OsStr>>,
+        ) -> Result<()> {
+            let all_targets: Option<&OsStr> = all_targets.then_some("--all-targets".as_ref());
+            let sh = Shell::new()?;
+            cmd!(sh, "cargo {toolchain} check {extra_flags...} --manifest-path {path} {all_targets...} {args...}").run()?;
+            Ok(())
+        }
+        // Check, and let caller control flags.
+        let toolchain = self.get_toolchain_override();
+        let miri_manifest = path!(self.miri_dir / "Cargo.toml");
+        let cargo_miri_manifest = path!(self.miri_dir / "cargo-miri" / "Cargo.toml");
+        check_package(miri_manifest.as_ref(), &toolchain, &self.cargo_extra_flags, true, flags)?;
+        check_package(
+            cargo_miri_manifest.as_ref(),
+            &toolchain,
+            &self.cargo_extra_flags,
+            false,
+            flags,
+        )?;
+        Ok(())
+    }
+
+    fn test(&mut self, bless: bool, flags: &[OsString]) -> Result<()> {
+        let miri_manifest = path!(self.miri_dir / "Cargo.toml");
+        let toolchain = self.get_toolchain_override();
+        // First build and get a sysroot.
+        Self::build_package(
+            miri_manifest.as_ref(),
+            &toolchain,
+            &self.cargo_extra_flags,
```

Comment:
If you make this a method (`self.build_package`), you don't have to pass `toolchain` and `cargo_extra_flags`.